### ISSUE TITLE
JAWS goes in and out of forms mode when there are adjacent, multiple, buttons and controls containing the aria-haspopup attribute.

### ISSUE THREAD

**Issue by accdc** [FreedomScientific/standards-support#82]
JAWS goes in and out of forms mode when there are adjacent, multiple, buttons and controls containing the aria-haspopup attribute.

JAWS  (v18, v2018) 
Firefox (v60, ESR), IE11
Windows7, Windows8, Windows10.

[auto-forms-mode-test.zip](https://github.com/FreedomScientific/VFO-standards-support/files/1992633/auto-forms-mode-test.zip)

```
//begin code snippet// 
<button aria-haspopup="dialog">button 1</button>
<button aria-haspopup="true">button 2</button> 
<button aria-haspopup="true">button 3</button> 
<button aria-haspopup="listbox">button 4</button>
<button aria-haspopup="tree">button 5</button>
<button aria-haspopup="grid">button 6</button>
//end code snippet//
```

To recreate: Tab to the buttons containing the aria-haspopup property. When you tab to the first button, JAWS goes into forms mode. When you tab to the second button, JAWS goes out of forms mode. When you tab to the third button, JAWS goes back into forms mode.

JAWS remains in forms mode if a subsequent control in the tab order contains the aria-haspopup property (with exception to aria-haspopup="dialog" - see entry #2, below -)

So, in the case of aria-haspopup opening a menu, when the button or link receives focus, it should be announced but standard navigation must be maintained. However, when this triggering element is activated and focus is then moved to an element within a Menu that includes a valid ARIA Menu role attribute for that widget construct, then Forms Mode should be automatically activated onFocus. The same would be true if focus is moved into an ARIA Listbox, Tree, Grid, or any other interactive widget role. This would ensure that the presence of aria-haspopup does not inadvertently activate automatic Forms Mode when this is not desired by the screen reader user, and it equally ensures that automatic Forms Mode will always be activated when focus is moved into interactive widgets that require keyboard navigation keystrokes to be passed directly to that widget. This is equally true for other supporting ARIA attributes like aria-pressed and aria-expanded.

As such, automatic Forms Mode should be activated whenever focus is set to an element that includes one of the following roles:

role="searchbox"
http://whatsock.com/training/matrices/#searchbox

role="scrollbar"
http://whatsock.com/training/matrices/#scrollbar

role="slider"
http://whatsock.com/training/matrices/#slider

role="spinbutton"
http://whatsock.com/training/matrices/#spinbutton

role="textbox"
http://whatsock.com/training/matrices/#textbox

role="combobox"
http://whatsock.com/training/matrices/#combobox

role="listbox" or role="option"
http://whatsock.com/training/matrices/#listbox

role="menu", role="menubar", role="menuitem", role="menuitemcheckbox", or role="menuitemradio"
http://whatsock.com/training/matrices/#menu
http://whatsock.com/training/matrices/#menubar

role="radiogroup" or role="radio"
http://whatsock.com/training/matrices/#radiogroup

role="tablist" or role="tab"
http://whatsock.com/training/matrices/#tablist

role="tree", role="group", or role="treeitem"
http://whatsock.com/training/matrices/#tree

role="grid", role="treegrid", role="row", role="gridcell", role="columnheader", or role="rowheader"
http://whatsock.com/training/matrices/#grid
http://whatsock.com/training/matrices/#treegrid

Additionally, automatic Forms Mode / Applications Mode should always be activated when focus is set to any active element within a container element including role="application", or when the element with role="application" receives focus."

**Comment by stevefaulkner** [comment]
[test case](https://s.codepen.io/stevef/debug/NEwpex)

**Comment by stevefaulkner** [comment]
Tested with JAWS 2019.1801using [test case](https://s.codepen.io/stevef/debug/NEwpex)

- With Internet Explorer 11 when first button tabbed to goes into forms mode, stays in forms mode for all buttons.
- With Firefox 64.0b10 (64-bit) when first button tabbed to goes into forms mode, out of forms mode  for buttons 2 & 3, back into forms mode for 4 to 6. Appaears that `aria-haspoup=true` on `button` results in exiting forms mode.

**Comment by stevefaulkner** [comment]
**Note:** in IE all the `button`s in the test case are exposed as `role=push button`. In firefox `button`s 2 & 3 are exposed as `role=menu button`, which accounts for the change in behaviour.

**Comment by stevefaulkner** [comment]
Note: adding explicit role=button to the aria-haspopup="true" button appears to stop JAWS going out of forms mode in Firefox:
` <button role="button" aria-haspopup="true">button  role="button" + aria-haspopup="true"</button>`

added above to test case

**Comment by accdc** [comment]
Hi,
I was asked by a client to check on the status of these bugs.

Do you know if the intent is to always turn Forms Mode on when focus is set to a button with aria-haspopup, or is there differing logic based on the value of the attribute?

The inconsistency doesn't make any sense to me.

Thanks,
Bryan

**Comment by JAWS-test** [comment]
I think the error is partly due to a Core AAM [specification error](https://github.com/w3c/core-aam/issues/51) that treats aria-haspopup differently for buttons than for all the other elements.

**Comment by jmegginson** [comment]
This ticket may be (now) obsolete with JAWS 2023 support of aria-haspopup=dialog.

I tested adjacent buttons with a combination of aria-haspopup=true and aria-haspopup=dialog per Bryan's example above. 

JAWS goes into forms mode (automatically) when a button contains aria-haspopup=true" which seems logical. 
Conversely JAWS refrains (or exits) form mode when an object with aria-haspopup=dialog is focused upon.

I was an advocate of this ticket in 2018 but now this appears to be working correctly with JAWS 2023.
